lib/src/default_index/revset_engine.rs

```diff
@@ -45,26 +45,36 @@ trait ToPredicateFn: fmt::Debug {
     /// Creates function that tests if the given entry is included in the set.
     ///
     /// The predicate function is evaluated in order of `RevsetIterator`.
-    fn to_predicate_fn(&self) -> BoxedPredicateFn<'_>;
+    fn to_predicate_fn<'a>(&self) -> BoxedPredicateFn<'a>
+    where
+        Self: 'a;
```

Comment:
Doesn't this still say that `Self` must outlive `'a`? I thought that's what `'_` meant too. At the end of this series, is the goal that the function returned from `containing_fn()` is supposed to outlive the revset? So the following should be allowed?

```rust
let contains = revset.containing_fn();
drop(revset);
contains(&commit_id);
```

I'm not sure I've understood the goal correctly.

Reply:
I read it as the `Self` type (such as fields in `Self`) outlives `'a`, so `BoxedPredicateFn<'a>` basically has the same lifetime as `Self`. I think @Ralith had more detailed/exact explanation in discord.

`'_` implies `<'a>(&'a self) -> 'a`, meaning the returned value can borrow `&self`.

> So the following should be allowed?
> 
> ```rust
> let contains = revset.containing_fn();
> drop(revset);
> contains(&commit_id);
> ```

Yes. Templater needs this because `contains` will be cached at `'repo` lifetime.

Reply:
But if `contains` in my example above has the same lifetime as `revset`, then we can't use `contains` after we've dropped `revset`. Oh, it doesn't mean that `Self` outlives `'a`; it means that any lifetimes on `Self` outlive `'a`. So if there are no lifetime bounds, then the lifetime is `'static`. https://doc.rust-lang.org/reference/trait-bounds.html#lifetime-bounds